src/api/eventTypes.js

```diff
@@ -100,23 +100,54 @@ export type PresenceEvent = $ReadOnly<{|
 |}>;
 
 /**
- * Updates the user status for a user
+ * New value for a user's self-assigned availability and/or text/emoji
+ *   status
  *
- * @prop [away] - update user's away status:
- *       - `true` the user is away regardless of presence
- *       - `false` remove the away status, now use presence
- * @prop [status_text] - if present:
- *       - empty string clears the user's status text
- *       - any string sets user's status to that
+ * See InitialDataUserStatus for the corresponding initial data. The
+ * per-user properties have different meanings between here and there.
  *
- * Not providing a property means 'leave this value unchanged'
  */
 export type UserStatusEvent = $ReadOnly<{|
   ...EventCommon,
   type: typeof EventTypes.user_status,
   user_id: UserId,
+
+  // Much of the below is observed empirically; we should get it into
+  // the doc:
+  //   https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Emoji.20statuses.20in.20zulip.2Eyaml/near/1322329
+
+  // true/false: User unavailable/available.
+  // omitted: User did not change this; keep current value
   away?: boolean,
+
+  // "foo": User's status text is "foo".
+  // "": User's status text is unset.
+  // omitted: User did not change this; keep current value
   status_text?: string,
+
+  // For these three emoji-status properties:
+  //
+  // non-empty string: User's status emoji is as indicated.
+  // "": User's status emoji is unset
+  // omitted: User did not change this; keep current value, or:
+  // TODO(server-5.0): Absent before feature level 86.
+  //
+  // "As indicated" means what the doc says, or:
+  //
+  // emoji_name is the emoji's human-readable name; then, for
+  //
+  // Unicode emoji:
+  //   reaction_type is 'unicode_emoji'
+  //   emoji_code is the emoji's Unicode codepoint
+  // realm emoji:
+  //   reaction_type is 'realm_emoji'
+  //   emoji_code is the ID in the realm-emoji namespace
+  // Zulip extra emoji:
+  //   reaction_type is 'zulip_extra_emoji'
+  //   emoji_code is the ID in the Zulip extra emoji namespace
```

Comment:
I think this is all covered in the API docs (following the links from this event's doc to https://zulip.com/api/add-reaction#parameters ), so we don't need to repeat it here.

It would be good to include a pointer to the `Reaction` type, though, because that includes some information that isn't in the docs. And, hmm, one thing the docs say:
> (`emoji_code` will be its Unicode codepoint).

is not true, and is incomplete even in the cases where it's partly or arguably true.

The truth is that a Unicode emoji may consist of a sequence of several code points, and there's a particular idiosyncratic way they have to be encoded to come up with the Zulip `emoji_code`.

The jsdoc on our `ReactionType` accurately says that a Unicode emoji is a sequence of code points. The jsdoc on our `Reaction#emoji_code` adds a link to the implementation, where a comment (which I wrote, back in 2018, after reverse-engineering this) explains the actual encoding -- here's a direct link:
https://github.com/zulip/zulip/blob/73266350efa3b6a39dc36744e91782010f974481/zerver/models.py#L2903-L2907

At the time I wrote that, the API docs didn't exist. That information should probably make it back into the API docs now that they do exist, but in the meantime we can point at the existing accurate version in our comments.

Reply:
I think actually the server doc should not link to any reaction-related docs. (I think it was a mistake to call one of these `reaction_type`.)

The right abstraction is "emoji", I think, not "reaction". It may be feasible for the server doc to deduplicate on an "emoji" concept—but I think it would be a fine improvement on the status quo for the doc to just spell out what is meant by each of these three "status emoji" properties.

Reply:
I agree that it's confusing to have "reaction" appear here, and is a sign that this area of the API isn't factored the right way.

I think it's not a coincidence that these three properties exactly coincide between status emoji and emoji reactions, though.

In reactions, the name `reaction_type`, and the use of the word "emoji" in each of the three existing values for that type, indicate that the idea when that was designed was that maybe there would be other kinds of "reactions" other than emoji, and the reactions would cover those. In that vision, it'd be a temporary coincidence that the structure of status emoji exactly coincides with the structure of reactions -- the wild and crazy unknown future concept of another "reaction" type that we might add might not make sense for user status.

But I think that early vision isn't happening. It's been years, and we have no feature ideas in which we'd want to overload "reactions" with something that isn't an emoji. If we did someday add a fourth value of `reaction_type`, I think it would basically still have to be some kind of emoji -- that's really what the UX of the reactions feature calls for. Something that truly doesn't fit as an emoji isn't going to fit in that reactions UX either, and so will use its own new bit of API surface.

That means that the structure we had for describing a "message reaction" is really all about an "emoji reaction" -- which means it's really just describing an emoji, plus saying what message to attach it to and on whose behalf. It just has this one quirk where one of the properties is named `reaction_type` and should really be `emoji_type`.

Given that, when there's this other feature that wants to describe an emoji, it's useful for the API to avoid multiplying concepts and reuse the "describe an emoji" concept it already has. Especially so given how gosh-darn *complicated* that concept is, as described at my `models.py` link above. That raises the cost (for the user of the API) of duplicating all that detail.

I think my ideal version of this status API would have called the property `emoji_type` instead of `reaction_type`. Then these three properties would exactly correspond to a generic "emoji" concept; and the doc on message reactions could point to here, or to a factored-out page elsewhere, and just mention that they have a quirk that they say `reaction_type` instead of `emoji_type`.

Reply:
Ah OK, yeah, that's reasonable.